### PR TITLE
Subsystem versioned

### LINKED ISSUE #909: Fix parsing/generation of installed dune files
The current behaviour in master regarding installed dune files is not right as they are parsed using the new lexer. That means that if a user upgrades from jbuilder to dune, dune will fail when trying to read an installed dune file installed by jbuilder.

## Parsing these files

Installed dune files start with: `(dune <version>`, where `<version>` is currently 1. Let's do the following to load these files:

- we manually fetch 3 tokens using `Sexp.Lexer.token` and check that they are `Lparen, Atom "dune", Atom s`. If `s` is `1` we parse the rest of the file using `Lexer.jbuild_token` and set `Stanza.syntax` to `(0, 0)`.
- if it is `2` we parse the rest of the file using `Lexer.token` and set `Stanza.syntax` to `(1, 0)`

### Versions of subsystems

Currently these files look like:

```
$ cat `ocamlfind query ppx_inline_test`/*.dune
(dune
 1
 ((inline_tests.backend
   1.0
   ((runner_libraries (ppx_inline_test.runner.lib))
    (flags
     (inline-test-runner
      ${library-name}
      -source-tree-root
      ${ROOT}
      -diff-cmd
      -))
    (generate_runner (echo "let () = Ppx_inline_test_lib.Runtime.exit ();;"))))))
```

Here the `1.0` was supposed to be the version of the sub-system. However, this doesn't correspond to the right thing anymore as the version of the inline test and ppx subsystems are aligned with the version of the dune language.

I have a proposal in mind where installed dune files will become similar to `dune-project` files, but it will be more work, so for now we should simply ignore this `1.0`.

## Generating these files

The syntax into which these files are generated should match the version of the file where the library is defined. So if the library was defined in a `jbuild` file, we should generate the file in version `1` (so jbuild syntax) if if the library was defined in a `dune` file, we should generate the file in version `2` (so dune syntax).

### PR DESCRIPTION
**Pull request by rgrinberg**
Fix #909 

This is still WIP because I should really be saving the syntax version rather than the Jbuild vs. Dune version in the Info files. But this PR is still useful for initial review.

One thing that I'm wondering about is we should still generate the per subsystem versions in syntax `(1, 0)`. Seems like jbuilder will not be able to load .dune files generated by dune anyway because of that initial 2.0 language.

### CODE REVIEW

**Comment by diml** [comment]
> One thing that I'm wdering about is we should still generate the per subsystem versions in syntax (1, 0). Seems like jbuilder will not be able to load .dune files generated by dune anyway because of that initial 2.0 language.

When the sub-system is defined in a jbuild file, dune should generate `dune 1` files. Otherwise it's not going to work for the released versions of ppx_inline_test, ppx_expect, ppxlib, etc...

**Comment by diml** [comment]
Actually I see that this is what is implemented by this PR

File: src/installed_dune_file.ml [hunk 1]
```diff
@@ -21,33 +21,54 @@ let parse_sub_systems sexps =
...
-  Sexp.Of_sexp.parse of_sexp Univ_map.empty (Io.Sexp.load ~mode:Single fname)
+  Io.with_lexbuf_from_file fname ~f:(fun lexbuf ->
+    let (version_loc, version) =
+      let bad_dune_file = "Unable to read (dune x.y ..) line file" in
+      let rec loop = function
+        | [_; _; _] as a -> List.rev a
+        | acc ->
+          begin match (Sexp.Lexer.token lexbuf : Sexp.Lexer.Token.t) with
+          | Eof ->
+            Loc.fail (Loc.in_file (Path.to_string fname)) "%s" bad_dune_file
```
Comment (diml) [inline src/installed_dune_file.ml]: You can return `List.rev acc`, the error will be caught by the `match loop [] with ...`.

File: src/install_rules.ml [hunk 2]
```diff
@@ -39,8 +39,9 @@ module Gen(P : Install_params) = struct
   let gen_lib_dune_file lib =
     SC.add_rule sctx
       (Build.arr (fun () ->
+         let lang = Option.value_exn (Lib.defined_using_lang lib) in
          Format.asprintf "%a@." (Sexp.pp Dune)
```
Comment (diml) [inline src/install_rules.ml]: s/Dune/lang/

File: src/installed_dune_file.ml [hunk 3]
```diff
@@ -21,33 +21,54 @@ let parse_sub_systems sexps =
...
+    in
+    let (lexer, syntax) =
+      match version with
+      | "1" -> (Sexp.Lexer.jbuild_token, (0, 0))
+      | "2" -> (Sexp.Lexer.token, (1, 0))
+      | _   -> Loc.fail version_loc "unknown version %S" version
+    in
+    Sexp.Of_sexp.parse of_sexp
+      (Univ_map.singleton (Syntax.key Stanza.syntax) syntax)
+      (Io.Sexp.load ~lexer ~mode:Single fname))
```
Comment (diml) [inline src/installed_dune_file.ml]: We shouldn't reopen the file a second time here. We need to construct a `lexer` function that first return the three extracted token and then switch to the real lexer.

**Comment by diml** [comment]
> This is still WIP because I should really be saving the syntax version rather than the Jbuild vs. Dune version in the Info files.

That seems right.

BTW, thinking about this again, where we read:

```
(dune
 2
 ((inline_tests.backend
   X.Y
```

we could set the `M.syntax` value of the subsystem to version `X.Y` since this is what we do when we generate the file.

**Comment by diml** [comment]
And we set it to `(0, 0)` when the file start with `(dune 1 ...`

**Comment by rgrinberg** [comment]
Jérémie Dimino <notifications@github.com> writes:

> When the sub-system is defined in a jbuild file, dune should generate
> `dune 1` files. Otherwise it's not going to work for the released
> versions of ppx_inline_test, ppx_expect, ppxlib, etc...

Agreed. The simplification would apply to subsystems in dene files.

File: src/installed_dune_file.ml (hunk 3 above)
Comment (rgrinberg) [reply src/installed_dune_file.ml]: I'm not really sure how to do it the way you say however. Seems like we'd need to call parser with an initial set of lexemes (the 3 lexemes that we already parsed).

What should be possible is just reading the file into a string and then constructing the lexer from_string a second time. That should cut down on the sys calls.

Comment (diml) [reply src/installed_dune_file.ml]: There are two ways of doing it:

first one:

```ocaml
let pending_tokens = ref ... in
let lexer lb =
  match !pending_tokens with
  | [] -> lexer lb
  | x :: l -> pending_tokens := l; x
in
...
```

second one, which implement all the logic in one function:

```ocaml
let state = ref 0 in
let lexer = ref !Sexp.Lexer.token in
let lexer lb =
  let token = !lexer lb in
  (match !state, token with
   | 0, Lparen -> state := 1
   | 1, Atom (A "dune") -> state := 2
   | 2, Atom (A "1") -> state := 3; lexer := Sexp.Lexer.jbuild_token
   | 2, Atom (A "2") -> state := 3;
   | 2, Atom (A s) -> error ()
   | 3, _ -> ()
   | _ -> error ());
  token
in
```

**Comment by diml** [comment]
BTW, should we just save the whole parsing context in libraries?

**Comment by rgrinberg** [comment]
Do you see other uses for it? I'd rather have Jbuild.Library.t abstract away the parsing related details kind of like Lib.Info.t abstracts away the source of libraries when making a Lib.t.

File: src/installed_dune_file.ml (hunk 3 above)
Comment (rgrinberg) [reply src/installed_dune_file.ml]: I see. I went for the first approach. Did not it was allowed to reuse the same state between possibly different lexer functions like this.

**Comment by rgrinberg** [comment]
> we could set the M.syntax value of the subsystem to version X.Y since this is what we do when we generate the file.

Sorry, this might be obtuse but isn't this done already? The X.Y in your example is a sub system version.

**Comment by rgrinberg** [comment]
I see, thanks for clarifiyng. Why only for dune 2 files however? We don't hide the syntax version when parsing other sexps that comes from jbuild files.

**Comment by diml** [comment]
> I see, thanks for clarifiyng. Why only for dune 2 files however? We don't hide the syntax version when parsing other sexps that comes from jbuild files.

Because in dune 1 files the version of sub-systems is set to 1.0, and `M.syntax` is `Stanza.syntax` for the inline tests and ppx driver subsystems, so we'd be setting `Stanza.syntax` to 1.0 instead of 0.0.

**Comment by rgrinberg** [comment]
@diml I tried doing this in my last commit. I admit that I'm not fully confident about my change however.

**Comment by diml** [comment]
That looks fine, we just a comment in the code explaining why we do this

**Comment by rgrinberg** [comment]
@diml added a comment

File: src/installed_dune_file.ml [hunk 4]
```diff
@@ -59,6 +98,12 @@ let gen confs =
   in
   Sexp.List
     [ Sexp.unsafe_atom_of_string "dune"
-    ; Sexp.unsafe_atom_of_string "1"
+    ; Sexp.unsafe_atom_of_string
+        (match dune_version with
+         | (0, 0) -> "1"
+         | (1, 0) -> "2"
+         | _ ->
```
Comment (diml) [inline src/installed_dune_file.ml]: Not sure we should be that strict. In practice, it's likely that we'll be increasing the version of the dune language often. In version 2 files, the exact version is saved in each sub-system, so we can accept any version >= 1.0 here.

File: src/stanza.ml [hunk 5]
```diff
@@ -13,7 +13,11 @@ let syntax =
     ]
 
 module File_kind = struct
-  type t = Jbuild | Dune
+  type t = Sexp.syntax = Jbuild | Dune
+
+  let of_syntax = function
+    | (0, 0) -> Jbuild
```
Comment (diml) [inline src/stanza.ml]: Even though we are not going to release another version of jbuilder, we should write `(0, _)` here (it's slightly clearer IMO)

File: src/installed_dune_file.ml [hunk 6]
```diff
@@ -38,52 +38,50 @@ let of_sexp =
...
-    let lexer =
-      let pending_tokens = ref tokens in
-      fun lb ->
-        match !pending_tokens with
-        | [] -> lexer lb
-        | x :: xs -> pending_tokens := xs; x
-    in
-    Sexp.Of_sexp.parse of_sexp
-      (Univ_map.singleton (Syntax.key Stanza.syntax) syntax)
+    Sexp.Of_sexp.parse of_sexp Univ_map.empty
```
Comment (rgrinberg) [inline src/installed_dune_file.ml]: How come you aren't passing the version in the parsing context?

**Comment by rgrinberg** [comment]
@diml the last commit relax the version checks I had before.

**Pull request merged**